src/eval_context.rs

```diff
@@ -669,6 +670,26 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
                         bug!("attempted to take a reference to an enum downcast lvalue"),
                 };
 
+                // Check alignment and non-NULLness.
+                let (_, align) = self.size_and_align_of_dst(ty, val)?;
+                if ptr.is_ptr() {
```

Comment:
or actually... match on the primval and get the inner parts. In the else path you do a `to_u128()` and in the main path a `to_ptr()`, I think a match would be clearer here.